### ISSUE TITLE
UI gets corrupted after using Factor normally, cross-platform rendering bug of some kind

### ISSUE THREAD

**Issue by erg** [factor/factor#1978]
After normal use, the UI gets corrupted in ways that look like this screenshot. If there is a bug open already, please dup this. Also, I am not 100% sure it happens on Linux, but please confirm it does.

![corrupt](https://user-images.githubusercontent.com/27430/37560890-7c36e13a-2a0e-11e8-8044-ce1833a72087.png)

**Comment by AlexIljin** [comment]
Apart from me complaining on the mailing list, I don't think we have a bug registered so far.
Here's a relevant video that I made two years ago. That stuff still happens, but randomly: sometimes after a long session, sometimes almost immediately. In the video you can see that the vertical scroll bar is not present, so the work session was very short (and I only ever open one Listener window).

**Comment by AlexIljin** [comment]
https://www.youtube.com/watch?v=uqdERZRTJ4Y

**Comment by AlexIljin** [comment]
This one time I tried digging into the contents of the Toolbar when this issue happened. I hoped there was some corruption of the cached rendered strings, but those were fine as far as I could tell.

**Comment by catb0t** [comment]
Have never ever seen this on Linux that I can recall.

**Comment by AlexIljin** [comment]
It doesn't only affect the Toolbar. It seems like some string pieces (a character or a repeated substring) gets substituted for a different one - throughout the Listener scrollable contents. I couldn't find the issue in the **cache,** but it sure seems like it's got to be there or about.
The Toolbar may be most affected because it's the oldest contents of the window, or something.

**Comment by mrjbq7** [comment]
More likely OpenGL texture corruption somehow like rebinding a texture ID over one that already existed for a different string. Had a buggy GL function call in minesweeper produce a similar result. But have not seen this corruption in normal use.

**Comment by AlexIljin** [comment]
After updating to the latest Factor build, this issue seems to be happening more often than before, and quicker from the start of the working session. Also, Factor application seems to randomly hang after mundane operations like starting a search in the help browser, running a unit-test, reloading vocabs with `F2` or starting a mouse selection in the listener.

**Comment by AlexIljin** [comment]
Before the latest update of my Factor application I worked with the binaries from `factor-windows-x86-64-2018-03-03-20-07.zip`. Recently I downloaded `factor-windows-x86-64-2018-07-10-02-10.zip` and extracted from there factor.exe, *.com, *.dll, and updated the sources from the GitHub repo. But the sources don't contribute to the issue: the hanging and the UI corruption happen even if I never use `F2`, but simply run from the downloaded image file.

**Comment by AlexIljin** [comment]
Here's something I wrote in the mailing list on 26.04.2016, I want it to be here for the record, as even I have trouble remembering the details:

"
  I've been seeing this kind of issues on at least three different PCs (Windows XP 32-bit and Windows 8.1 64-bit). Both in the most recent stable build of Factor, and in several recent development builds.

  The issue seems to be a kind of memory corruption. It usually happens after I have the Listener open for a long time (days) or after a short time of memory-intensive work. There is no reliable way to reproduce the issue, and the corruption is random in the specific gadgets it affects. The toolbar is affected most often, but other contents of the Listener can have switched texts. You work for some time, and suddenly screen gadgets display the wrong text or character. Usually you can see where that text comes from, i.e. it's not completely random, but rather existing text pieces get mixed around and randomly redistributed between gadgets. If a gadget reacts to mouse interactions, you can usually see the previous, "normal" text, like in the video.
"

I'll add a Lenovo laptop Y70-70 with Windows 10 Pro x64 1703 to the list of systems where I had this issue. It's very hard to reproduce for me, too. Maybe I was most affected by it when I was developing my own graphical gadgets, or when I had memory-intensive database work, but in any case rerunning the same code didn't ever reproduce the corruption of the previous session that I had to terminate.

The Listener has its text contents displayed in "bursts", which is obvious when you select the text with mouse: some words or even phrases are selected as a unit. I think those correspond to individual `write` calls, or something like that. The corruption always works by substituting the contents of one such burst with a different one. The same burst will be replaced by the same substitute in all places on screen and throughout the history (if you scroll upwards). Selecting the text will typically show the original, correct unsubstituted text, and clearing the selection will restore the substitution/corruption. This makes me think that the OpenGL texture corruption/rebinding theory has some credibility to it. The question is, how do we reproduce that.

**Comment by catb0t** [comment]
Are they all Intel integrated graphics? and what kind of OpenGL support do they have built in?

**Comment by AlexIljin** [comment]
I'll try to reply to this in a few days, when I'm done traveling. Sorry for the delay.

**Comment by AlexIljin** [comment]
Another piece of information into this puzzle: the corruption seems to hit on a per-world basis. I was writing some unit-tests, and just now I had a corruption that replaced the strings `"1.txt"` with `}` in all of the Listener, but the same original string was unaffected in the Errors window.

Also, if you open a new Listener window, it doesn't inherit the corruption (but also it doesn't inherit the data stack and the command history), so it can be seen as a partial workaround.

Is there a way to flush the texture cache of a world?

**Comment by AlexIljin** [comment]
![2018-07-29 listener](https://user-images.githubusercontent.com/169039/43370266-31c38cca-937c-11e8-8b08-c3ec8ca1f5b4.png)
The interesting part of this fresh screenshot is that it's not the toolbar that got corrupted. Instead the toolbar's texts corrupted the other parts of the window. All I did was reload the vocab a few times and run its unit-tests a few times. Nothing out of the ordinary.

**Comment by erg** [comment]
Can you find/name/post a vocabulary that loading/running tests a few times reliably breaks it? I would love to fix this somehow...

**Comment by AlexIljin** [comment]
@erg: no such thing, I'm afraid. It's all random.
Even the code such as this doesn't reproduce the issue:
```
USE: tools.scaffold
"corrupt-listener" scaffold-work


USING: corrupt-listener formatting io.encodings.utf8 ;
"corrupt-listener" scaffold-tests

100 <iota> [
    dup
    "resource:work/corrupt-listener/corrupt-listener.factor" utf8 [
        ": word%d ( -- ? ) t ;\n" printf
    ] with-file-appender
    "resource:work/corrupt-listener/corrupt-listener-tests.factor" utf8 [
        "[ t ] [ word%d ] unit-test\n" printf
    ] with-file-appender
    "corrupt-listener" [ reload ] [ test ] bi
] each
```

There is a possibility that some user gestures are necessary (mouse clicks, selections, etc.).
I don't know if it's related, but in the released v0.98 I sometimes encounter a hard hang following a mouse click on a piece of code or a mouse selection attempt in the listener.

**Comment by AlexIljin** [comment]
The "hard hang" mentioned above got its own issue number: #2037.

**Comment by AlexIljin** [comment]
Recently I returned to working on my code, and had this issue multiple times with the latest available Windows build: Factor 0.99 x86.64 (1889, heads/master-fd7e19b6a0, Sep  7 2018 12:08:06).

At one point there was some interesting deviation. My code runs in the Listener and updates a label with the progress of an operation (pasted into text with `gadget.`). The text is refreshed up to 10 times per second (subject to the actual progress made by the algorithm). So, this dynamically changing text was not only appearing on the label, but also several other places on screen showed updates, seemingly taking turns: place 1 would get updated with the latest text, then place 2, then place 3. So, I didn't have 4 counters running in parallel, but one continuous counter (where I expected it to be), and three other lazy counters, updating out of sync with each other, showing different strings from the past and updating irregularly.

Does that give you any new ideas?

I have two video cards in my laptop: "Intel(R) HD Graphics 4600" and "NVIDIA GeForce GTX 960M". I don't know which one is used for Factor, but I would assume the first one.

**Comment by AlexIljin** [comment]
I tried making the screen recording of this bug, and I wanted to catch the new dynamic behavior, so I discarded some attempts that bugged statically (in the manner described at the start of this thread). I kept statistics of the run, and I can tell you that out of 24 runs I had 5 that bugged out statically (sometimes with a single word out of place on screen), and the 24th one was dynamic.

From the run I have the feeling that Factor needs to do something with memory to bug out like that. This is why I had previously suggested that the issue may be provoked by the Garbage Collector. More specifically, during this run I could not make the bug manifest itself reliably while I was simply running the app time after time. But when I switched to the browser to write the previous comment, the very next run produced the video you can enjoy by visiting the link below.

https://youtu.be/-Q3R_nsXTLA

The delay at the start of the `.verify-path` work is caused by Factor loading a 78Mb SQLite DB from the given root path. If you make repeated runs, it doesn't take as long as you see in the video, but after doing something else with the PC, it may be gone from the cache, or whatever, so it takes a long time. Maybe Factor does some harder than otherwise necessary work on the memory, like causing page misses, or something?

In any case, my best reproduction rate is 6/24, which is pretty annoying in real life, but not 100% consistent.

What do you think about all this?

**Comment by AlexIljin** [comment]
One more bit of information: if you start a new Listener with `A-L`, it will not inherit the problem, even though the process is the same. So, the corruption is limited to a world instance, which makes me think it's more of a Factor issue than an OpenGL issue.

**Comment by AlexIljin** [comment]
I'm going to try to instrument the `draw-world` and lower implementation details with `logging:add-logging` and see what the issue could be. Can anyone point out which vocabs constitute the drawing backend? Is it `windows.opengl32`, or something else?

While I have the time during the holidays I could try to understand how it all works at the lower level, track down the problem. I seem to remember there was a blog or a web-site of some sort that explained the basics of what OpenGL is and how it works. Could someone give me a link to such a resource?

**Comment by AlexIljin** [comment]
My first action would be to wrap the `update-ui-loop` call in `start-ui-thread` with `with-logging`.

**Comment by AlexIljin** [comment]
Just to confirm that the issue is not specific to my graphics card and/or its drivers, I have reproduced the same issue in a VirtualBox VM with Windows 7 Pro. The result was similar to the screenshot in the first post here, and it took me only a few minutes of normal work in the Listener/Watcher.

**Comment by erg** [comment]
I also hit it recently. We could write some more tools/tooling to catch ui errors maybe?

**Comment by AlexIljin** [comment]
@erg, that would be great.

I analyzed the last Youtube video that I posted here. Watching it frame by frame, I noticed that updating the controls with the wrong data goes in a loop through the same sequence of controls over and over, and the length of the loop is about 10 items. Whenever there is a change on the screen (a new number is output to the label, or the text caret blinks), the next control in the loop gets updated with the previous text of the label.

In general, the text that's output to the screen is cached as pre-rendered `single-texture` and `script-string` tuples. It is purged from the cache after it has stayed there for 10 world updates. See `cache:purge-cache` (the default value for `max-age` is 10). The purging is done in-place with `assoc-filter!`

I suspect there might be sort of concurrency issue with updating the texture cache. For my further experiments I'm going to try setting the `max-age` to 5 in `ui.text:world-text-handle`, and see if it affects the output loop.

**Comment by AlexIljin** [comment]
I've pushed the `ui-instrumenting` branch to my repo, if you want to see what I did so far.

**Comment by melted** [comment]
A simple experiment could be to remove the 2cache call in draw-string and see if the problem goes away.

**Comment by AlexIljin** [comment]
I tried removing `2cache` from `draw-string`, and in a short time factor.exe ate all the available memory and hanged. With every blink of the cursor it allocated 2+ Mb from the OS while its windows were visible (I had the Listener and the Browser open).

**Comment by AlexIljin** [comment]
BTW, some simple experiments with the Process Explorer open alongside factor.exe show that the blinking of the cursor forces the full world to be redrawn. This is very inefficient. The cursors always treated specially by the OSs for that very reason, and factor UI should, too.

**Comment by mrjbq7** [comment]
Instead of 2caching it, just call dispose (which is what the cache does when it gets freed up, so you don’t run out of memory).

**Comment by AlexIljin** [comment]
@mrjbq7, thanks for the tip about disposing the texture, I'll certainly do that.

Is there a good reason why GC doesn't support calling a finalization method? When a piece of memory is garbage-collected, why isn't there a `destroy` method to be called for all objects that chose to implement it?

**Comment by catb0t** [comment]
@AlexIljin it's not automatic i don't think, but there is [`with-destructors`](https://docs.factorcode.org/content/word-with-destructors,destructors.html)

**Comment by AlexIljin** [comment]
@catb0t, thank you for the suggestion.

In XDS Modula/Oberon compiler there was a system method called `finalize`, which any class could override, and which would be called by the GC for each object of that class before its disposal. I remember there was a warning in the documentation about the method being called in the context of the GC with a limited call stack space available. And, of course, one was not to create new pointers to the object being destroyed inside that method. It was a perfect mechanism for automatically disposing of external resources, such as file or image handles allocated from the OS, and it allowed one to create weak links.

I read in the documentation that Factor went the other way, and we are supposed to manually wrap everything with `with-destructors`, which seems a little too low-level and hands-on. Is there a compelling reason not to have the automated finalization?

**Comment by AlexIljin** [comment]
Running the same program that gave me issues recorded on the last video with the `2cache` replaced with `with-disposal`, I no longer see the GUI corruption problem. I'll run some more tests to make sure it's really gone, since it's an issue that's not 100% reproducible.

**Comment by AlexIljin** [comment]
Can it be that caching sometimes caches the wrong thing? Like, associating a string with a different (previously created) texture?

**Comment by mrjbq7** [comment]
I’ve been somewhat convinced that the OpenGL texture cache per world is getting mixed up, but I don’t have any evidence except your reports on Windows and once or twice in a blue moon on macOS. 

That’s a good piece of info! And maybe a workaround for you temporarily.

**Comment by mrjbq7** [comment]
The OpenGL texture cache is a numeric ID to a texture that’s drawn with an OpenGL call using the ID, I think. 

So either two worlds that are using the same GL context (collisions) or maybe something not freeing properly. 

Out of curiosity, do you have this issue with a single Factor window open? Or only when more than one is open?

**Comment by mrjbq7** [comment]
We do need to improve the caching algorithm also, max age by number of accesses is a little bit not intelligent.  

Maybe the purge cache is clearing something it shouldn’t, or vice verse.

**Comment by AlexIljin** [comment]
All corruptions I've seen this far are limited to one window, no matter how many windows are open by the process. The contents of one window gets mixed up inside it, I've seen no spilling of contents between windows. Most of the time it's limited to the (single) Listener window, even if there is a Browser or other windows open by the same Factor process.

I'll keep running my tests in two parallel Factor instances, one with only the Listener, and one with also the Browser open.

**Comment by AlexIljin** [comment]
Unfortunately, despite running several Factor instances in parallel I could not reproduce the issue neither in the instance with texture caching disabled, nor in the unmodified version. Sometimes this bug is really annoying, and sometimes it's impossible to reproduce.

**Comment by melted** [comment]
The effect of this bug can be replicated by taking a texture from the cache and dispose it (I did `world get world-text-handle` and inspected it, and pushed one of the cached textures to the listener, and then called dispose on it, it can then be fixed by deleting it from the cache). So my best guess is that some other code has disposed a bunch of textures that still remains in the cache and then when retrieved they show the texture that OpenGL has reused the id for.

A band aid would be to check if the texture has been disposed when retrieving from the cache (or making a cache that know about disposables and drop them if they have been disposed). But that might mask the real bug.

**Comment by melted** [comment]
There isn't any other code that touches those textures, they only exist in the cache, so that can't be the right explanation. But I know a bit what to look for now the next time I see this.

**Comment by melted** [comment]
The cache seems incredibly quick in discarding stuff, if I scroll up and down, it only ever contains on-screen text.

**Comment by mrjbq7** [comment]
Good investigations!

I’ve noticed the less-than-ideal cache behavior as well.  It’s a kind of time-to-live cache and expires based on number of accesses. Probably we should allow older objects as long as it doesn’t grow beyond some threshold of max size. Ideas welcome!

If you can “reproduce” the behavior by disposing a texture that is still cached, that’s useful to direct the hunt. I haven’t seen this on macOS so I’m unsure if it’s platform-specific (Windows?) or affects every platform. That might implicate different code.

**Comment by melted** [comment]
I finally managed to catch it in a corrupted state. And the textures that have gone bad are marked as disposed. So it's at least possible to throw the bad ones away when retrieving them from the cache. I'll see if I can find the root cause for the disposal.

**Comment by melted** [comment]
At least we can say that it's Factor thing, not a system thing, if OpenGL had a bug that deleted the textures in the background, the disposed flag wouldn't have been set. But I have the hardest time finding anywhere the bug can be. The textures only live in the assoc, the assoc is only accessed from a few places, and a whole bunch of them gets disposed at the same time. The two places that disposes the entries are in `purge-cache` and `clear-assoc`. `clear-assoc` is only called via `dispose` when the window is closed.

That leaves `purge-cache`, called in `draw-world` but that filters the entries it disposes fairly definitely. 

Add that this only occurs very occasionally and it's kind of hard to see a mechanism.

**Comment by melted** [comment]
Hmmm, what happens when the underlying hashtable grows?

**Comment by AlexIljin** [comment]
Is it possible that some of the textures are marked with `|dispose`, and then an error is thrown inside a `with-destructors` call? The error may be handled and suppressed, but the texture may still be disposed and remain in the cache.

Can a disposed (on purpose) texture remain in the assoc by mistake (via another pointer, for example).

Did you know that there are not only `string` objects in the assoc, but also `fonts:selection`s? I don't know the lifecycle of the latter, when they are created and when destroyed. Anyway, those were not involved when I had the chances to reproduce this issue.

Could we have some special code paths playing role when the assoc or parts of it are relocated in memory by the GC?

I wish to add that this situation is most likely a consequence of manual memory management. I insist that destructors should be called automatically when the object is collected by the GC, and not via some `with-destructors` scoping. The manual approach is very easy to do wrong, very difficult to test for correctness in all use cases, and very difficult to track down in situations like this one.

**Comment by AlexIljin** [comment]
Could you make a special mechanism similar to the `continuation` slot for `disposable`, but one that would store the continuation when the object was disposed, not allocated. If this mechanism could be activated only for the `texture` class, we could see where it was wrongfully disposed.

**Comment by melted** [comment]
Good idea, I can just reuse the continuation slot and fill it on disposal.

**Comment by melted** [comment]
The GC can't be trusted to ever delete objects, so it's not useful for freeing resources in general, which might be arbitrarily scarce, and it obviously wouldn't be deterministic. It would also require a finalizer mechanism, and that will impact the performance of the GC. Java has recently removed finalizers, because it stood in the way of making performant collectors.

A reference count mechanism would be interesting for this use case if there was a way to automatically changing the count when creating and dropping references.

**Comment by AlexIljin** [comment]
Why is it that in the whole codebase `<cache-assoc>` is only followed by `&dispose` in `"windows.uniscribe" add-startup-hook`, and not in any other similar mechanism (not in `ui.text.pango` and not in `core-text`)?

**Comment by mrjbq7** [comment]
Hmm, thats weird we use ``&dispose`` but not inside a ``with-destructors``... that seems like a bug @AlexIljin.  Fixed in bd2ed2c2148ae8e94ee7742993981bda2e883092.  But probably we should dispose in a shutdown-hook, or maybe it doesn't matter?

Do you mostly see problems on Windows?  Maybe that's the root cause?

**Comment by melted** [comment]
I don't think it's the root case for this. I added the saving of the continuation and the disposes comes from purge-cache, so I put in code that serialized the cache before the purge and saved that and the state after if a disposed texture was still in the cache. From that I learned that it started when texture was correctly disposed but not deleted. And then it cascaded subsequently. I'd like to improve my tracing code, but it looks like it's operating on a stale hashtable so the deletions doesn't get back into the cache. I didn't find a realistic way that could happen and I haven't tried recreate it since then. I want to instrument it some more before I do.

It seems to require input from the user, it's not possible to recreate just by dumping a lot of text in the listener.

**Comment by erg** [comment]
Have we seen this since https://github.com/factor/factor/commit/bd2ed2c2148ae8e94ee7742993981bda2e883092?

**Comment by mrjbq7** [comment]
I think you reverted my change with your master patch couple days ago...

It shouldn’t have fixed the bug though.

**Comment by mrjbq7** [comment]
I re-submitted my patch.  Let's see if it resolves any of the windows ui
corruption issues.

**Comment by erg** [comment]
Oops, sorry!

**Comment by mrjbq7** [comment]
It should probably be the way it was with &dispose since that adds to the global destructors list that’s cleaned up on shutdown. I made the patch as a test to see if we are accidentally running those when we shouldn’t and corrupting the UI.
